events/suite_test.go

```diff
@@ -0,0 +1,138 @@
+package events_test
+
+import (
+	"context"
+	"fmt"
+	"testing"
+	"time"
+
+	"github.com/awslabs/operatorpkg/events"
+	"github.com/awslabs/operatorpkg/object"
+	"github.com/awslabs/operatorpkg/test"
+	. "github.com/awslabs/operatorpkg/test/expectations"
+	"github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+	prometheus "github.com/prometheus/client_model/go"
+	"github.com/samber/lo"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes/scheme"
+	clock "k8s.io/utils/clock/testing"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/client/fake"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/metrics"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+var ctx context.Context
+var fakeClock *clock.FakeClock
+var controller *events.Controller[*test.TestObject]
+var kubeClient client.Client
+
+func Test(t *testing.T) {
+	lo.Must0(test.SchemeBuilder.AddToScheme(scheme.Scheme))
+	RegisterFailHandler(Fail)
+	RunSpecs(t, "Events")
+}
+
+var _ = BeforeSuite(func() {
+	fakeClock = clock.NewFakeClock(time.Now())
+	kubeClient = fake.NewClientBuilder().WithScheme(scheme.Scheme).WithIndex(&corev1.Event{}, "involvedObject.kind", func(o client.Object) []string {
+		evt := o.(*corev1.Event)
+		return []string{evt.InvolvedObject.Kind}
+	}).Build()
+	controller = events.NewController[*test.TestObject](kubeClient, fakeClock)
+	ctx = log.IntoContext(context.Background(), ginkgo.GinkgoLogr)
+})
+
+var _ = Describe("Controller", func() {
+	BeforeEach(func() {
+		controller.EventCount.Reset()
+	})
+	It("should emit metrics on an event", func() {
+		events := []*corev1.Event{}
+		for i := range 5 {
+			events = append(events, createEvent("test-object", fmt.Sprintf("Test-type-%d", i), fmt.Sprintf("Test-reason-%d", i)))
+			ExpectApplied(ctx, kubeClient, events[i])
+		}
+
+		for i := range 5 {
+			Expect(getMetric("operator_testobject_event_total", conditionLabels(fmt.Sprintf("Test-type-%d", i), fmt.Sprintf("Test-reason-%d", i)))).To(BeNil())
+		}
+
+		for i := range 5 {
+			_, err := reconcile.AsReconciler(kubeClient, controller).Reconcile(ctx, reconcile.Request{NamespacedName: client.ObjectKeyFromObject(events[i])})
+			Expect(err).ToNot(HaveOccurred())
+		}
+
+		for i := range 5 {
+			Expect(getMetric("operator_testobject_event_total", conditionLabels(fmt.Sprintf("Test-type-%d", i), fmt.Sprintf("Test-reason-%d", i))).GetCounter().GetValue()).To(BeEquivalentTo(1))
```

Comment:
Can these not happen in the same loop?

Reply:
I wanted each set to happen one at a time. First creating all the events, reconciling all the values and then validating each one. The idea being that we can check if multiple  event reconciliation was happening. I'm committed to that idea, so I can take in your comment

Reply:
If you'd like to test multiple event reconciliation, would a waitgroup make sense? https://github.com/aws/karpenter-provider-aws/blob/7e8269bc5b333afdd6863d9e7b933a5bc049cd2f/test/suites/scale/deprovisioning_test.go#L183